lua/modules/diffusion/diffusion.cc

```diff
@@ -2,19 +2,162 @@
 // SPDX-License-Identifier: MIT
 
 #include "lua/framework/console/console.h"
+#include "framework/parameters/input_parameters.h"
 #include "framework/runtime.h"
 #include "framework/physics/solver_base/solver.h"
-#include "modules/cfem_diffusion/cfem_diffusion_solver.h"
-#include "lua/modules/cfem_diffusion/cfem_diff_solver.h"
-#include "modules/dfem_diffusion/dfem_diffusion_solver.h"
-#include "lua/modules/dfem_diffusion/dfem_diff_solver.h"
+#include "modules/diffusion/cfem_diffusion_solver.h"
+#include "modules/diffusion/dfem_diffusion_solver.h"
+#include "modules/diffusion/fv_diffusion_solver.h"
+#include "modules/diffusion/mg_diffusion_solver.h"
+#include "lua/framework/math/functions/lua_scalar_spatial_material_function.h"
 
 namespace opensnlua
 {
 
 namespace
 {
 
+std::shared_ptr<LuaScalarSpatialMaterialFunction>
+CreateFunction(const std::string& function_name)
+{
+  opensn::ParameterBlock blk;
+  blk.AddParameter("lua_function_name", function_name);
+  opensn::InputParameters params = LuaScalarSpatialMaterialFunction::GetInputParameters();
+  params.AssignParameters(blk);
+  return std::make_shared<LuaScalarSpatialMaterialFunction>(params);
+}
+
+} // namespace
+
+// CFEM diffusion solver
+
+opensn::ParameterBlock
+CFEMSetOptions(const opensn::InputParameters& params)
+{
+  const std::string fname = __FUNCTION__;
+
+  params.RequireParameter("arg0");
+  params.RequireParameter("arg1");
+
+  auto d_coef_function = CreateFunction("D_coef");
+  opensn::function_stack.push_back(d_coef_function);
+
+  auto q_ext_function = CreateFunction("Q_ext");
+  opensn::function_stack.push_back(q_ext_function);
+
+  auto sigma_a_function = CreateFunction("Sigma_a");
+  opensn::function_stack.push_back(sigma_a_function);
+
+  const size_t handle = params.GetParamValue<size_t>("arg0");
+  auto& solver =
+    opensn::GetStackItem<opensn::diffusion::CFEMSolver>(opensn::object_stack, handle, fname);
+  solver.SetDCoefFunction(d_coef_function);
+  solver.SetQExtFunction(q_ext_function);
+  solver.SetSigmaAFunction(sigma_a_function);
+
+  auto options_params = opensn::diffusion::CFEMSolver::OptionsBlock();
+  options_params.AssignParameters(params.GetParam("arg1"));
+
+  solver.SetOptions(options_params);
```

Comment:
Is there a base diffusion solver that could house this since it seems to be the same for all diffusion solvers?

Reply:
There is not yet. I think I would like to see the tests for MG and FV solvers first. 

It is obvious that some base class for the diffusion solver is needed. There is a lot of overlap ATM. But, one step at a time :-)

Reply:
That would be really helpful 👍 Thanks!

Reply:
Yep, not asking for anything now! At some point, there was a single diffusion solver that had discretization options (i.e. if you used CFEM it called a different assembly routine than DFEM). Something like that could be worth thinking about in the future

Reply:
I suppose this is not what you are looking for: mip_diffusion_3d_3a_dsa_ortho.lua

Reply:
> I suppose this is not what you are looking for: mip_diffusion_3d_3a_dsa_ortho.lua

No. That's using the DFEM solver which was already converted. As far as I could tell, there were no tests for the MG or FV solver.